src/meta/src/hummock/compaction/level_selector.rs

```diff
@@ -226,23 +233,61 @@ impl DynamicLevelSelector {
 }
 
 impl LevelSelector for DynamicLevelSelector {
-    fn need_compaction(&self, levels: &[Level], level_handlers: &mut [LevelHandler]) -> bool {
-        let ctx = self.get_priority_levels(levels, level_handlers);
-        ctx.score_levels
-            .first()
-            .map(|(score, _)| *score > SCORE_BASE)
-            .unwrap_or(false)
+    fn need_compaction(
+        &self,
+        levels: &[Level],
+        level_handlers: &mut [LevelHandler],
+        hash_mapping_manager: Option<&HashMappingManager>,
+    ) -> bool {
+        if hash_mapping_manager.is_some() {
+            true
+        } else {
+            let ctx = self.get_priority_levels(levels, level_handlers);
+            ctx.score_levels
+                .first()
+                .map(|(score, _)| *score > SCORE_BASE)
+                .unwrap_or(false)
+        }
     }
 
     fn pick_compaction(
         &self,
         task_id: u64,
         levels: &[Level],
         level_handlers: &mut [LevelHandler],
+        hash_mapping_manager: Option<&HashMappingManager>,
     ) -> Option<SearchResult> {
         let ctx = self.get_priority_levels(levels, level_handlers);
         for (score, level_idx) in ctx.score_levels {
             if score <= SCORE_BASE {
+                if let Some(hash_mapping_manager) = hash_mapping_manager {
```

Comment:
If score is less than SCORE_BASE, it means that there is no level need compaction

Reply:
If you want to always check the whole data, please move this code to an independent picker such as `HashMappingCompactionPicker`.  And you can make a new selector such as `HashMappingSelector`.
```rust
struct HashMappingSelector {
    hash_mapping_manager: HashMappingManagerRef,
}

impl LevelSelector for HashMappingSelector {
    fn need_compaction(
        &self,
        levels: &[Level],
        level_handlers: &mut [LevelHandler],
    ) -> bool { .. }
    fn pick_compaction(
        &self,
        task_id: u64,
        levels: &[Level],
        level_handlers: &mut [LevelHandler],
    ) -> Option<SearchResult> { .. }
}

pub struct CompactStatus {
    pub(crate) level_handlers: Vec<LevelHandler>,
    pub(crate) next_compact_task_id: u64,
    compaction_selectors: Vec<Box<dyn LevelSelector>>,
}

impl CompactStatus {
     pub fn new() -> Self {
         compaction_selectors.push(Box::new(DynamicLevelSelector::new(config, overlap_strategy)));
         compaction_selectors.push(Box::new(HashMappingSelector::new(hash_mapping_manager));
         Self {
              compaction_selectors,
              ..
          }
     }
}

```

I think if you check every bitmap in every sstable it will take a long time when the number of files increase. It seems that every table-id will have it's own map-hash value, so we can calling this function when the hash-value

Reply:
> If you want to always check the whole data, please move this code to an independent picker such as `HashMappingCompactionPicker`. And you can make a new selector such as `HashMappingSelector`.
> 
> ```rust
> struct HashMappingSelector {
>     hash_mapping_manager: HashMappingManagerRef,
> }
> 
> impl LevelSelector for HashMappingSelector {
>     fn need_compaction(
>         &self,
>         levels: &[Level],
>         level_handlers: &mut [LevelHandler],
>     ) -> bool { .. }
>     fn pick_compaction(
>         &self,
>         task_id: u64,
>         levels: &[Level],
>         level_handlers: &mut [LevelHandler],
>     ) -> Option<SearchResult> { .. }
> }
> 
> pub struct CompactStatus {
>     pub(crate) level_handlers: Vec<LevelHandler>,
>     pub(crate) next_compact_task_id: u64,
>     compaction_selectors: Vec<Box<dyn LevelSelector>>,
> }
> 
> impl CompactStatus {
>      pub fn new() -> Self {
>          compaction_selectors.push(Box::new(DynamicLevelSelector::new(config, overlap_strategy)));
>          compaction_selectors.push(Box::new(HashMappingSelector::new(hash_mapping_manager));
>          Self {
>               compaction_selectors,
>               ..
>           }
>      }
> }
> ```

I will do that after other parts of this PR being resolved